File: languages/java/src/ads/TestRunner.java
package ads;

import ads.gen.TestData;
import java.util.List;

/**
 * Runs every catalog vector against the Java implementations and asserts the
 * expected output. Exit status 0 == all pass; 1 == any failure.
 */
public final class TestRunner {

    private static int failures;
    private static int passed;

    private TestRunner() {
    }

    public static void main(String[] args) {
        runAlgorithms();
        runStructures();
        System.out.printf("Java: %d passed, %d failed.%n", passed, failures);
        if (failures > 0) {
            System.exit(1);
        }
    }

    // ---- parsing helpers ----------------------------------------------------

    private static int[] parseInts(String token) {
        if (token == null || token.isBlank()) {
            return new int[0];
        }
        String[] parts = token.trim().split("\\s+");
        int[] out = new int[parts.length];
        for (int i = 0; i < parts.length; i++) {
            out[i] = Integer.parseInt(parts[i]);
        }
        return out;
    }

    private static List<List<Integer>> parseGraph(String token) {
        return Solutions.parseGraph(token);
    }

    // ---- algorithm dispatch -------------------------------------------------

    private static String[] ids = {
        "two_sum", "binary_search", "merge_sort", "quick_sort",
        "max_subarray", "lcs", "knapsack_01", "edit_distance",
        "graph_bfs", "graph_dfs"
    };

    private static void runAlgorithms() {
        String[][][] data = TestData.ALGORITHMS;
        for (int p = 0; p < ids.length; p++) {
            for (String[] caseVec : data[p]) {
                check(ids[p], caseVec[0], caseVec[1]);
            }
        }
    }

    private static void check(String id, String input, String expected) {
        String[] parts = input.split("\\|", -1);
        try {
            switch (id) {
                case "two_sum": {
                    int[] got = Solutions.twoSum(parseInts(parts[0]), Integer.parseInt(parts[1]));
                    checkIntArr(id, input, expected, got, parseInts(expected));
                    break;
                }
                case "binary_search": {
                    int got = Solutions.binarySearch(parseInts(parts[0]), Integer.parseInt(parts[1]));
                    checkInt(id, input, expected, got, Integer.parseInt(expected));
                    break;
                }
                case "merge_sort":
                case "quick_sort": {
                    int[] got = id.equals("merge_sort")
                            ? Solutions.mergeSort(parseInts(input))
                            : Solutions.quickSort(parseInts(input));
                    checkIntArr(id, input, expected, got, parseInts(expected));
                    break;
                }
                case "max_subarray": {
                    int got = Solutions.maxSubarray(parseInts(parts[0]));
                    checkInt(id, input, expected, got, Integer.parseInt(expected));
                    break;
                }
                case "lcs": {
                    int got = Solutions.lcs(parts[0], parts[1]);
                    checkInt(id, input, expected, got, Integer.parseInt(expected));
                    break;
                }
                case "knapsack_01": {
                    int got = Solutions.knapsack01(
                            Integer.parseInt(parts[0]), parseInts(parts[1]), parseInts(parts[2]));
                    checkInt(id, input, expected, got, Integer.parseInt(expected));
                    break;
                }
                case "edit_distance": {
                    int got = Solutions.editDistance(parts[0], parts[1]);
                    checkInt(id, input, expected, got, Integer.parseInt(expected));
                    break;
                }
                case "graph_bfs": {
                    int got = Solutions.graphBfs(
                            parseGraph(parts[0]), Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
                    checkInt(id, input, expected, got, Integer.parseInt(expected));
                    break;
                }
                case "graph_dfs": {
                    boolean got = Solutions.graphDfs(
                            parseGraph(parts[0]), Integer.parseInt(parts[1]), Integer.parseInt(parts[2]));
                    boolean want = expected.equals("true");
                    if (got != want) {
                        recordFailure(id, input, String.valueOf(got), expected);
                    } else {
                        passed++;
                    }
                    break;
                }
                default:
                    throw new IllegalStateException("unhandled " + id);
            }
        } catch (RuntimeException e) {
            recordFailure(id, input, "<exception>", expected);
        }
    }

    private static void checkInt(String id, String input, String expected, int got, int want) {
        if (got != want) {
            recordFailure(id, input, String.valueOf(got), expected);
        } else {
            passed++;
        }
    }

    private static void checkIntArr(
            String id, String input, String expected, int[] got, int[] want) {
        if (got.length != want.length) {
            recordFailure(id, input, java.util.Arrays.toString(got), expected);
            return;
        }
        for (int i = 0; i < got.length; i++) {
            if (got[i] != want[i]) {
                recordFailure(id, input, java.util.Arrays.toString(got), expected);
                return;
            }
        }
        passed++;
    }

    private static void recordFailure(String id, String input, String got, String expected) {
        failures++;
        System.out.printf(
                "FAIL %s input=%s -> got %s, expected %s%n", id, input, got, expected);
    }

    // ---- structures ----------------------------------------------------------

    private static String[] structIds = {
        "stack", "queue", "linked_list", "bst", "trie", "min_heap"
    };

    private static void runStructures() {
        String[][][] data = TestData.STRUCTURES;
        for (int i = 0; i < structIds.length; i++) {
            for (String[] caseVec : data[i]) {
                checkOps(structIds[i], caseVec[0]);
            }
        }
    }

    private static void checkOps(String id, String encoded) {
        Object instance = newInstance(id);
        for (String token : encoded.split(",", -1)) {
            if (token.isBlank()) {
                continue;
            }
            String[] parts = token.split(":", -1);
            String op = parts[0];
            String arg = "";
            String expected = "";
            if (voidOps().contains(op)) {
                arg = parts.length > 1 ? parts[1] : "";
            } else if (resultOps().contains(op)) {
                expected = parts.length > 1 ? parts[1] : "";
            } else {
                arg = parts.length > 1 ? parts[1] : "";
                expected = parts.length > 2 ? parts[2] : "";
            }
            applyOp(instance, op, arg, expected);
        }
    }

    private static java.util.Set<String> voidOps() {
        return java.util.Set.of("push", "enqueue", "append", "prepend", "insert");
    }

    private static java.util.Set<String> resultOps() {
        return java.util.Set.of(
                "pop", "dequeue", "peek", "size", "is_empty",
                "min", "max", "height", "to_list", "in_order");
    }

    private static Object newInstance(String id) {
        switch (id) {
            case "stack":
                return new Structures.Stack();
            case "queue":
                return new Structures.Queue();
            case "linked_list":
                return new Structures.LinkedList();
            case "bst":
                return new Structures.BST();
            case "trie":
                return new Structures.Trie();
            case "min_heap":
                return new Structures.MinHeap();
            default:
                throw new IllegalStateException("unhandled structure " + id);
        }
    }

    private static void applyOp(Object instance, String op, String arg, String expected) {
        if (instance instanceof Structures.Stack stack) {
            applyStack(stack, op, arg, expected);
        } else if (instance instanceof Structures.Queue queue) {
            applyQueue(queue, op, arg, expected);
        } else if (instance instanceof Structures.LinkedList list) {
            applyList(list, op, arg, expected);
        } else if (instance instanceof Structures.BST bst) {
            applyBst(bst, op, arg, expected);
        } else if (instance instanceof Structures.Trie trie) {
            applyTrie(trie, op, arg, expected);
        } else if (instance instanceof Structures.MinHeap heap) {
            applyHeap(heap, op, arg, expected);
        } else {
            throw new IllegalStateException("no handler for " + instance.getClass());
        }
    }

    private static boolean hasExpected(String expected) {
        return !expected.isEmpty();
    }

    private static void applyStack(Structures.Stack stack, String op, String arg, String expected) {
        switch (op) {
            case "push":
                stack.push(Integer.parseInt(arg));
                break;
            case "pop":
                expect(result(stack.pop()), expected, "stack.pop");
                break;
            case "peek":
                expect(result(stack.peek()), expected, "stack.peek");
                break;
            case "is_empty":
                expect(stack.isEmpty() ? "true" : "false", expected, "stack.is_empty");
                break;
            case "size":
                expect(result(stack.size()), expected, "stack.size");
                break;
            default:
                throw new IllegalStateException("stack op " + op);
        }
    }

    private static void applyQueue(Structures.Queue queue, String op, String arg, String expected) {
        switch (op) {
            case "enqueue":
                queue.enqueue(Integer.parseInt(arg));
                break;
            case "dequeue":
                expect(result(queue.dequeue()), expected, "queue.dequeue");
                break;
            case "peek":
                expect(result(queue.peek()), expected, "queue.peek");
                break;
            case "is_empty":
                expect(queue.isEmpty() ? "true" : "false", expected, "queue.is_empty");
                break;
            case "size":
                expect(result(queue.size()), expected, "queue.size");
                break;
            default:
                throw new IllegalStateException("queue op " + op);
        }
    }

    private static void applyList(Structures.LinkedList list, String op, String arg, String expected) {
        switch (op) {
            case "append":
                list.append(Integer.parseInt(arg));
                break;
            case "prepend":
                list.prepend(Integer.parseInt(arg));
                break;
            case "get":
                expect(result(list.get(Integer.parseInt(arg))), expected, "list.get");
                break;
            case "contains":
                expect(list.contains(Integer.parseInt(arg)) ? "true" : "false", expected, "list.contains");
                break;
            case "remove":
                expect(list.remove(Integer.parseInt(arg)) ? "true" : "false", expected, "list.remove");
                break;
            case "size":
                expect(result(list.size()), expected, "list.size");
                break;
            case "to_list":
                expectes(ListToStr(list.toList()), expected, "list.to_list");
                break;
            default:
                throw new IllegalStateException("list op " + op);
        }
    }

    private static void applyBst(Structures.BST bst, String op, String arg, String expected) {
        switch (op) {
            case "insert":
                bst.insert(Integer.parseInt(arg));
                break;
            case "contains":
                expect(bst.contains(Integer.parseInt(arg)) ? "true" : "false", expected, "bst.contains");
                break;
            case "remove":
                expect(bst.remove(Integer.parseInt(arg)) ? "true" : "false", expected, "bst.remove");
                break;
            case "min":
                expect(result(bst.min()), expected, "bst.min");
                break;
            case "max":
                expect(result(bst.max()), expected, "bst.max");
                break;
            case "height":
                expect(result(bst.height()), expected, "bst.height");
                break;
            case "size":
                expect(result(bst.size()), expected, "bst.size");
                break;
            case "in_order":
                expectes(ListToStr(bst.inOrder()), expected, "bst.in_order");
                break;
            default:
                throw new IllegalStateException("bst op " + op);
        }
    }

    private static void applyTrie(Structures.Trie trie, String op, String arg, String expected) {
        switch (op) {
            case "insert":
                trie.insert(arg);
                break;
            case "search":
                expect(trie.search(arg) ? "true" : "false", expected, "trie.search");
                break;
            case "starts_with":
                expect(trie.startsWith(arg) ? "true" : "false", expected, "trie.starts_with");
                break;
            case "size":
                expect(result(trie.size()), expected, "trie.size");
                break;
            default:
                throw new IllegalStateException("trie op " + op);
        }
    }

    private static void applyHeap(Structures.MinHeap heap, String op, String arg, String expected) {
        switch (op) {
            case "push":
                heap.push(Integer.parseInt(arg));
                break;
            case "pop":
                expect(result(heap.pop()), expected, "heap.pop");
                break;
            case "peek":
                expect(result(heap.peek()), expected, "heap.peek");
                break;
            case "is_empty":
                expect(heap.isEmpty() ? "true" : "false", expected, "heap.is_empty");
                break;
            case "size":
                expect(result(heap.size()), expected, "heap.size");
                break;
            default:
                throw new IllegalStateException("heap op " + op);
        }
    }

    private static String result(int value) {
        return String.valueOf(value);
    }

    private static String ListToStr(List<Integer> values) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < values.size(); i++) {
            if (i > 0) {
                sb.append(' ');
            }
            sb.append(values.get(i));
        }
        return sb.toString();
    }

    private static void expect(String got, String expected, String label) {
        if (got.equals(expected)) {
            passed++;
        } else {
            recordFailure(label, "<op>", got, expected);
        }
    }

    private static void expectes(String got, String expected, String label) {
        expect(got.trim(), expected.trim(), label);
    }
}

File: languages/cpp/tests/test_runner.cpp
// Cross-language catalog test runner (C++17).
// Reads generated test vectors, runs each through ads::solutions / ads::structures.

#include <cstdlib>
#include <iostream>
#include <set>
#include <sstream>
#include <string>
#include <vector>

#include "generated_test_data.hpp"
#include "ads/solutions.hpp"
#include "ads/structures.hpp"

namespace {

int failures = 0;
int passed = 0;

void report_fail(const std::string& label, const std::string& input,
                 const std::string& got, const std::string& expected) {
    ++failures;
    std::cerr << "FAIL " << label << " input=" << input << " -> got " << got
              << ", expected " << expected << '\n';
}

std::vector<int> parse_ints(const std::string& token) {
    std::vector<int> out;
    std::istringstream stream(token);
    int value = 0;
    while (stream >> value) {
        out.push_back(value);
    }
    return out;
}

std::vector<std::string> split(const std::string& text, char delimiter) {
    std::vector<std::string> parts;
    std::string current;
    std::istringstream stream(text);
    while (std::getline(stream, current, delimiter)) {
        parts.push_back(current);
    }
    return parts;
}

std::vector<std::vector<int>> parse_graph(const std::string& encoded) {
    std::vector<std::vector<int>> graph;
    if (encoded.empty()) {
        return graph;
    }
    for (const std::string& row : split(encoded, ';')) {
        auto pair = split(row, ':');
        if (pair.empty()) {
            continue;
        }
        graph.push_back(pair.size() > 1 ? parse_ints(pair[1])
                                        : std::vector<int>{});
    }
    return graph;
}

void check_int(const std::string& id, const std::string& input, int got,
               int expectedish) {
    if (got != expectedish) {
        report_fail(id, input, std::to_string(got), std::to_string(expectedish));
    } else {
        ++passed;
    }
}

void check_int_vector(const std::string& id, const std::string& input,
                      const std::vector<int>& got,
                      const std::vector<int>& expected) {
    if (got != expected) {
        std::ostringstream a;
        std::ostringstream b;
        for (size_t i = 0; i < got.size(); ++i) {
            if (i > 0) a << ' ';
            a << got[i];
        }
        for (size_t i = 0; i < expected.size(); ++i) {
            if (i > 0) b << ' ';
            b << expected[i];
        }
        report_fail(id, input, a.str(), b.str());
    } else {
        ++passed;
    }
}

void check_algorithms() {
    const auto& data = ads::gen::ALGORITHMS;
    const std::vector<std::string> ids = {
        "two_sum",       "binary_search", "merge_sort",     "quick_sort",
        "max_subarray",  "lcs",           "knapsack_01",    "edit_distance",
        "graph_bfs",     "graph_dfs",
    };
    for (size_t p = 0; p < ids.size(); ++p) {
        for (const auto& testcase : data[p]) {
            const std::string& input = testcase[0];
            const std::string& expected = testcase[1];
            auto parts = split(input, '|');
            const std::string& id = ids[p];
            if (id == "two_sum") {
                auto nums = parse_ints(parts[0]);
                int target = std::atoi(parts[1].c_str());
                check_int_vector(id, input, ads::two_sum(nums, target),
                                 parse_ints(expected));
            } else if (id == "binary_search") {
                int got = ads::binary_search(parse_ints(parts[0]),
                                             std::atoi(parts[1].c_str()));
                check_int(id, input, got, std::atoi(expected.c_str()));
            } else if (id == "merge_sort" || id == "quick_sort") {
                auto nums = parse_ints(input);
                auto got = id == "merge_sort" ? ads::merge_sort(nums)
                                              : ads::quick_sort(nums);
                check_int_vector(id, input, got, parse_ints(expected));
            } else if (id == "max_subarray") {
                check_int(id, input, ads::max_subarray(parse_ints(input)),
                          std::atoi(expected.c_str()));
            } else if (id == "lcs") {
                check_int(id, input, ads::lcs(parts[0], parts[1]),
                          std::atoi(expected.c_str()));
            } else if (id == "knapsack_01") {
                int got = ads::knapsack_01(std::atoi(parts[0].c_str()),
                                           parse_ints(parts[1]),
                                           parse_ints(parts[2]));
                check_int(id, input, got, std::atoi(expected.c_str()));
            } else if (id == "edit_distance") {
                check_int(id, input, ads::edit_distance(parts[0], parts[1]),
                          std::atoi(expected.c_str()));
            } else if (id == "graph_bfs") {
                int got = ads::graph_bfs(parse_graph(parts[0]),
                                         std::atoi(parts[1].c_str()),
                                         std::atoi(parts[2].c_str()));
                check_int(id, input, got, std::atoi(expected.c_str()));
            } else if (id == "graph_dfs") {
                bool got = ads::graph_dfs(parse_graph(parts[0]),
                                          std::atoi(parts[1].c_str()),
                                          std::atoi(parts[2].c_str()));
                bool want = expected == "true";
                if (got != want) {
                    report_fail(id, input, got ? "true" : "false", expected);
                } else {
                    ++passed;
                }
            }
        }
    }
}

const std::set<std::string> kVoidOps = {"push",  "enqueue", "append",
                                        "prepend", "insert"};
const std::set<std::string> kResultOps = {
    "pop",      "dequeue",       "peek",     "is_empty", "size",
    "min",      "max",           "height",   "to_list",  "in_order"};

void run_structure_case(const std::string& id, const std::string& encoded) {
    ads::Stack stack;
    ads::Queue queue;
    ads::LinkedList list;
    ads::BST bst;
    ads::Trie trie;
    ads::MinHeap heap;

    auto run = [&](const std::string& op, const std::string& arg,
                   const std::string& expected) {
        auto check = [&](const std::string& label, std::string got) {
            if (expected.empty()) {
                ++passed;
                return;
            }
            if (got == expected) {
                ++passed;
            } else {
                report_fail(label, id + ":" + op, got, expected);
            }
        };

        if (id == "stack") {
            if (op == "push") {
                stack.push(std::atoi(arg.c_str()));
                ++passed;
            } else if (op == "pop") {
                check("stack.pop", std::to_string(stack.pop()));
            } else if (op == "peek") {
                check("stack.peek", std::to_string(stack.peek()));
            } else if (op == "is_empty") {
                check("stack.is_empty", stack.is_empty() ? "true" : "false");
            } else if (op == "size") {
                check("stack.size", std::to_string(stack.size()));
            }
        } else if (id == "queue") {
            if (op == "enqueue") {
                queue.enqueue(std::atoi(arg.c_str()));
                ++passed;
            } else if (op == "dequeue") {
                check("queue.dequeue", std::to_string(queue.dequeue()));
            } else if (op == "peek") {
                check("queue.peek", std::to_string(queue.peek()));
            } else if (op == "is_empty") {
                check("queue.is_empty", queue.is_empty() ? "true" : "false");
            } else if (op == "size") {
                check("queue.size", std::to_string(queue.size()));
            }
        } else if (id == "linked_list") {
            if (op == "append") {
                list.append(std::atoi(arg.c_str()));
                ++passed;
            } else if (op == "prepend") {
                list.prepend(std::atoi(arg.c_str()));
                ++passed;
            } else if (op == "get") {
                check("list.get", std::to_string(list.get(std::atoi(arg.c_str()))));
            } else if (op == "contains") {
                check("list.contains",
                      list.contains(std::atoi(arg.c_str())) ? "true" : "false");
            } else if (op == "remove") {
                check("list.remove",
                      list.remove(std::atoi(arg.c_str())) ? "true" : "false");
            } else if (op == "size") {
                check("list.size", std::to_string(list.size()));
            } else if (op == "to_list") {
                std::ostringstream out;
                auto values = list.to_list();
                for (size_t i = 0; i < values.size(); ++i) {
                    if (i > 0) out << ' ';
                    out << values[i];
                }
                check("list.to_list", out.str());
            }
        } else if (id == "bst") {
            if (op == "insert") {
                bst.insert(std::atoi(arg.c_str()));
                ++passed;
            } else if (op == "contains") {
                check("bst.contains",
                      bst.contains(std::atoi(arg.c_str())) ? "true" : "false");
            } else if (op == "remove") {
                check("bst.remove",
                      bst.remove(std::atoi(arg.c_str())) ? "true" : "false");
            } else if (op == "min") {
                check("bst.min", std::to_string(bst.min()));
            } else if (op == "max") {
                check("bst.max", std::to_string(bst.max()));
            } else if (op == "height") {
                check("bst.height", std::to_string(bst.height()));
            } else if (op == "size") {
                check("bst.size", std::to_string(bst.size()));
            } else if (op == "in_order") {
                std::ostringstream out;
                auto values = bst.in_order();
                for (size_t i = 0; i < values.size(); ++i) {
                    if (i > 0) out << ' ';
                    out << values[i];
                }
                check("bst.in_order", out.str());
            }
        } else if (id == "trie") {
            if (op == "insert") {
                trie.insert(arg);
                ++passed;
            } else if (op == "search") {
                check("trie.search", trie.search(arg) ? "true" : "false");
            } else if (op == "starts_with") {
                check("trie.starts_with", trie.starts_with(arg) ? "true" : "false");
            } else if (op == "size") {
                check("trie.size", std::to_string(trie.size()));
            }
        } else if (id == "min_heap") {
            if (op == "push") {
                heap.push(std::atoi(arg.c_str()));
                ++passed;
            } else if (op == "pop") {
                check("heap.pop", std::to_string(heap.pop()));
            } else if (op == "peek") {
                check("heap.peek", std::to_string(heap.peek()));
            } else if (op == "is_empty") {
                check("heap.is_empty", heap.is_empty() ? "true" : "false");
            } else if (op == "size") {
                check("heap.size", std::to_string(heap.size()));
            }
        }
    };

    for (const std::string& token : split(encoded, ',')) {
        if (token.empty()) {
            continue;
        }
        auto parts = split(token, ':');
        const std::string& op = parts[0];
        std::string arg;
        std::string expected;
        if (kVoidOps.count(op)) {
            arg = parts.size() > 1 ? parts[1] : "";
        } else if (kResultOps.count(op)) {
            expected = parts.size() > 1 ? parts[1] : "";
        } else {
            arg = parts.size() > 1 ? parts[1] : "";
            expected = parts.size() > 2 ? parts[2] : "";
        }
        run(op, arg, expected);
    }
}

void check_structures() {
    const auto& ids = std::vector<std::string>{
        "stack", "queue", "linked_list", "bst", "trie", "min_heap"};
    const auto& data = ads::gen::STRUCTURES;
    for (size_t i = 0; i < ids.size(); ++i) {
        for (const auto& testcase : data[i]) {
            run_structure_case(ids[i], testcase[0]);
        }
    }
}

}  // namespace

int main() {
    check_algorithms();
    check_structures();
    std::cout << "C++: " << passed << " passed, " << failures << " failed.\n";
    return failures == 0 ? 0 : 1;
}